subsys/bluetooth/host/classic/l2cap_br.c

```diff
@@ -238,18 +240,71 @@ static uint8_t l2cap_br_get_ident(void)
 	return ident;
 }
 
+static void raise_data_ready(struct bt_l2cap_br_chan *br_chan)
+{
+	if (!atomic_set(&br_chan->_pdu_ready_lock, 1)) {
+		sys_slist_append(&br_chan->chan.conn->upper_data_ready,
+				 &br_chan->_pdu_ready);
+		LOG_DBG("data ready raised");
+	} else {
+		LOG_DBG("data ready already");
+	}
+
+	bt_conn_data_ready(br_chan->chan.conn);
+}
+
+static void lower_data_ready(struct bt_l2cap_br_chan *br_chan)
+{
+	struct bt_conn *conn = br_chan->chan.conn;
+	sys_snode_t *s = sys_slist_get(&conn->upper_data_ready);
+
+	__ASSERT_NO_MSG(s == &br_chan->_pdu_ready);
+	(void)s;
```

Comment:
can we fix `__ASSERT()` instead?

Reply:
Do we see the issue with unused variables when compiling ourselves, or is only Coverity?

Reply:
yes. I think it's clang builds in CI that complain about it. IMO `__ASSERT(x)` should expand to `(void)(x)`. But it seems it just expands to nothing.

Reply:
we do have `BT_ASSERT()`. I've for a long time thought about forcing `CONFIG_ASSERT=y` for bluetooth.
What we could do instead (if we all agree), is use `BT_ASSERT()` everywhere and `default y` in a non-settable kconfig.
If users want to `fast-car-with-no-brakes` they will have to go out of their way to disable it, but it would still be possible.
I think that's better than having `CONFIG_ASSERT` default to `n` and having to remember to turn it on for every app.

Reply:
>  I think that's better than having CONFIG_ASSERT default to n and having to remember to turn it on for every app.

Isn't the point of ASSERTS that they are off by default, and should only be enabled when debugging? 

> use BT_ASSERT() everywhere 

If we want to keep BT_ASSERT, then yes. I've always considered BT_ASSERT to be a thing that would be removed in the future, similar to BT_LOG, as it's just doing (more or less) the same but with a different prefix, right?

Reply:
> But instead of littering the code with if (verybad) {die();} I'd prefer having an assert that's always on for bluetooth.
My (our? with alwa) goal is that we can run Bluetooth in a userspace thread and then when verybad happens, we only kill that thread, allowing the app to recover and re-start bluetooth.

That sounds like something different than an assert. From https://en.wikipedia.org/wiki/Assertion_(software_development): 
> Assertions document logically impossible situations and discover programming errors

Impossible situations would indicate that it cannot be recovered, right? Sounds more like what you are looking for at (optional) runtime checks.

Reply:
nono, rare "illogical" bugs can happen.
That's why the softdevice controller ships with asserts on.
Whether it's the compiler that you can't trust (will reorder the code), or some charged particle flipping a bit, or a yet-undiscovered errata.

Asserts allow the controller to die, and to be re-started by the host.
Similarly, I'd like the host to be allowed to die, and be re-started by the app.

That's what I meant by recover. Not really recover the state of the host, but be able to wipe the state and start anew. The chance of that kind of condition triggering again is very slim.

Reply:
The wikipedia page is interesting

> In other cases, such as Java, assertions are present in the deployed code, and can be turned on in the field for debugging.[[2]](https://en.wikipedia.org/wiki/Assertion_(software_development)#cite_note-2)

> Assertions may also be used to promise the compiler that a given edge condition is not actually reachable, thereby permitting certain [optimizations](https://en.wikipedia.org/wiki/Compiler_optimizations) that would not otherwise be possible. In this case, disabling the assertions could actually reduce performance.

So enabling asserts actually make the code go faster 🏎️

Reply:
yep. but we don't have exceptions in C, "asserts" is what we have I guess.